### PR TITLE
[`flake8-type-checking`] Adds implementation for TC006

### PR DESCRIPTION
**Pull request by Daverball**
Now that `TCH` has been renamed to `TC` and the redirect from `TCH006` to `TCH010` can no longer bite us, I've added an implementation for this very simple rule.

## Summary

`TC006` checks for non-string literal arguments to `typing.cast` which adds unnecessary runtime overhead, since the function doesn't do anything at runtime.

This PR has a very tiny overlap with my other PR for `TCH007`/`TCH008` in `flake8_type_checking/helpers.rs` for adding the `quote_type_expression` function, but the two changes don't really depend on one another, so they can be merged in any order.

## Test Plan

`cargo nextest run`

### CODE REVIEW

**Comment by Daverball** [comment]
The extra hits for most of the other rules make sense to me, since fixing TC006 can result in subsequent TC001-003 errors.

I have no idea what the extra SIM115 violation is about though.

File: crates/ruff_linter/src/rules/flake8_type_checking/mod.rs [hunk 1]
```diff
@@ -62,6 +63,20 @@ mod tests {
...
+            Path::new("flake8_type_checking").join(path).as_path(),
+            &settings::LinterSettings {
+                preview: PreviewMode::Enabled,
+                ..settings::LinterSettings::for_rule(rule_code)
+            },
+        )?;
+        assert_messages!(snapshot, diagnostics);
+        Ok(())
+    }
+
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_type_checking/mod.rs:79]: I know this is a pattern that we have in other linter-groups too, but I don't think it's actually necessary to test a preview-rule. It's only necessary if the rule depends on any other preview-only behavior

File: crates/ruff_linter/src/rules/flake8_type_checking/rules/runtime_cast_value.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    if let Some(edit) = edit.as_ref() {
+        if checker
+            .comment_ranges()
+            .has_comments(type_expr, checker.source())
+        {
+            diagnostic.set_fix(Fix::unsafe_edit(edit.clone()));
+        } else {
+            diagnostic.set_fix(Fix::safe_edit(edit.clone()));
+        }
+    }
```
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/flake8_type_checking/rules/runtime_cast_value.rs:67]: The clone calls here seem unnecessary
```suggestion
    let edit = quote_type_expression(type_expr, checker.semantic(), checker.stylist()).ok();
    if let Some(edit) = edit {
        if checker
            .comment_ranges()
            .has_comments(type_expr, checker.source())
        {
            diagnostic.set_fix(Fix::unsafe_edit(edit));
        } else {
            diagnostic.set_fix(Fix::safe_edit(edit));
        }
    }
```

**Review by MichaReiser — APPROVED** [review]

**Comment by MichaReiser** [comment]
Thanks for contributing this rule. The code changes look good to me. 

I want to wait for updated ecosystem checks. I don't see why this PR should result in any new diagnostics other than TC006.

File: crates/ruff_linter/resources/test/fixtures/flake8_type_checking/TC006.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+def f():
+    import typing
+
+    typing.cast(int, 3.0)  # TC006
+
+
+def f():
+    import typing as t
+
+    t.cast(int, 3.0)  # TC006
```
Comment (MichaReiser) [inline crates/ruff_linter/resources/test/fixtures/flake8_type_checking/TC006.py:52]: Could we add a test where parts of the type are quoted?
```suggestion
    t.cast(Literal["3"] | Literal['3'], 3.0)  # TC006
```

**Comment by MichaReiser** [comment]
I'm a bit confused by the ecosystem checks. I understand that we run ruff exactly once and don't apply any fixes. That makes it unclear to me why there would be any changed diagnostics other than for TC006.

**Comment by Daverball** [comment]
> I'm a bit confused by the ecosystem checks. I understand that we run ruff exactly once and don't apply any fixes. That makes it unclear to me why there would be any changed diagnostics other than for TC006.

The only way this output makes sense to me is if it does in fact repeatedly apply fixes until it runs out fixes and uses the output from whatever the last iteration was for the ecosystem results. It's also a little sus that there's no +/- on the count of fixes, since these are definitely all auto-fixable. Maybe this was changed at some point without anyone noticing? Because the fixes count was always zero on my other PRs as well, even though they definitely included some new fixes.

**Comment by MichaReiser** [comment]
Something's definitely broken. Running the two ruff binaries without the ecosystem wrapper script gives me zero changes for pandas.

**Comment by Daverball** [comment]
> Something's definitely broken. Running the two ruff binaries without the ecosystem wrapper script gives me zero changes for pandas.

Indeed, I double checked, there's not even any type aliases defined in the first file in pandas. Maybe it's the result of broken caching on the CI side?

**Comment by MichaReiser** [comment]
I can reproduce the ecosystem differences locally....

**Comment by Daverball** [comment]
> I can reproduce the ecosystem differences locally.... I'm sure it's something obvious and very stupid :D

Actually I confused myself, this is about `typing.cast` and not type aliases. So it actually still makes sense. If you apply the fix for TC006, you will get the TC001 errors afterwards, because the file uses `from __future__ import annotations`, so the only runtime use of those `_typing` imports happened inside `cast` statements, so if you quote them, there are none left and `TC001-003` triggers.

So I still think ecosystem script applies fixes repeatedly and uses either the state where it can no longer fix things or the state just before it fixed everything for the results.

**Review by MichaReiser — APPROVED** [review]

**Comment by Daverball** [comment]
> Ohhhhh.... pandas sets `fix = true`

I see, ecosystem should probably force that to `false` then. Although I would see some value in checking for fix circularity issues in ecosystem, but that would require a larger restructure of how `ruff` is invoked and the results are interpreted.

**Pull request merged**